packages/nimble-components/src/rich-text/editor/tests/rich-text-editor-mention.spec.ts

```diff
@@ -0,0 +1,824 @@
+                [{ key: 'user:1', displayName: 'username1' }]
+            );
+            await appendTestMentionConfiguration(element, [
+                { key: 'test:2', displayName: 'testname2' }
+            ]);
+
+            expect(pageObject.getMarkdownRenderedTagNames()).toEqual([
+                'P',
+                RICH_TEXT_MENTION_USERS_VIEW_TAG,
+                RICH_TEXT_MENTION_USERS_VIEW_TAG,
+                'BR'
+            ]);
+            expect(
+                pageObject.getEditorMentionViewAttributeValues('mention-label')
+            ).toEqual(['username1', 'testname2']);
+            mappingElements[0]!.key = 'invalid';
+            await waitForUpdatesAsync();
+
+            expect(pageObject.getEditorTagNames()).toEqual(['P', 'A', 'A']);
+            expect(pageObject.getEditorLeafContents()).toEqual([
+                'user:1',
+                'test:2'
+            ]);
+        });
+    });
+
+    it('should fire "mention-update" event from configuration elment when there is @mention in editor', async () => {
+        const { userMentionElement } = await appendUserMentionConfiguration(
+            element
+        );
+        const mentionUpdateSpy = jasmine.createSpy('mention-update');
+        userMentionElement.addEventListener('mention-update', mentionUpdateSpy);
+        await pageObject.setEditorTextContent('@test');
+        expect(mentionUpdateSpy).toHaveBeenCalledTimes(1);
+    });
+
+    it('should fire "mention-update" event from configuration elment when there is update in @mention in editor', async () => {
+        const { userMentionElement } = await appendUserMentionConfiguration(
+            element
+        );
+        const mentionUpdateSpy = jasmine.createSpy('mention-update');
+        userMentionElement.addEventListener('mention-update', mentionUpdateSpy);
+        await pageObject.setEditorTextContent('@test');
+        expect(mentionUpdateSpy).toHaveBeenCalledTimes(1);
+    });
+
+    it('should fire "mention-update" event from configuration elment when pasting @ in editor', async () => {
+        const { userMentionElement } = await appendUserMentionConfiguration(
+            element
+        );
+        const mentionUpdateSpy = jasmine.createSpy('mention-update');
+        userMentionElement.addEventListener('mention-update', mentionUpdateSpy);
+        pageObject.pasteToEditor('@test');
+        await waitForUpdatesAsync();
+        expect(mentionUpdateSpy).toHaveBeenCalledTimes(1);
+    });
+
+    it('should fire "mention-update" event when deleting an existing @mention in editor', async () => {
+        await pageObject.setEditorTextContent('@test');
+        const { userMentionElement } = await appendUserMentionConfiguration(
+            element
+        );
+        const mentionUpdateSpy = jasmine.createSpy('mention-update');
+        userMentionElement.addEventListener('mention-update', mentionUpdateSpy);
+        await pageObject.sliceEditorContent(2);
```

Comment:
I don't think the implementation of `sliceEditorContent()` is correct for the intent of these tests. It basically behaves by manipulating the markdown in memory and then replacing the entire contents of the editor. This means these tests are more like paste tests. The test below which validates no event fired when `@` is removed is in fact validating that setting contents to a string with no `@` in it won't fire the event.

Instead I think `sliceEditorContent()` should set the focus at the end of the text and then press the Delete key a specified number of times. (The method might have a new name like `deleteEditorContent()`). Then these tests would be updated to assert that deleting the last character from `@test` does fire an event but deleting the last character from `@` doesn't fire an event.

Reply:
Sounds good. I manually verified that the events fire correctly now, so we'll have to rely on that until you implement the integration test approach.